examples/pytorch/graphsage/advanced/train_sampling_unsupervised_cache.py

```diff
@@ -0,0 +1,354 @@
+    g : The entire graph.
+    inputs : The features of all the nodes.
+    labels : The labels of all the nodes.
+    val_mask : A 0-1 mask indicating which nodes do we actually compute the accuracy for.
+    device : The GPU device to evaluate on.
+    """
+    model.eval()
+    with th.no_grad():
+        # single gpu
+        if isinstance(model, SAGE):
+            pred = model.inference(
+                g, nfeat, device, args.batch_size, args.num_workers
+            )
+        # multi gpu
+        else:
+            pred = model.module.inference(
+                g, nfeat, device, args.batch_size, args.num_workers
+            )
+    model.train()
+    return compute_acc(pred, labels, train_nids, val_nids, test_nids)
+
+
+#### Entry point
+def run(proc_id, n_gpus, args, devices, data):
+    # Unpack data
+    device = th.device(devices[proc_id])
+    if n_gpus > 0:
+        th.cuda.set_device(device)
+    if n_gpus > 1:
+        dist_init_method = "tcp://{master_ip}:{master_port}".format(
+            master_ip="127.0.0.1", master_port="12345"
+        )
+        world_size = n_gpus
+        th.distributed.init_process_group(
+            backend="nccl",
+            init_method=dist_init_method,
+            world_size=world_size,
+            rank=proc_id,
+        )
+    train_nid, val_nid, test_nid, n_classes, g, nfeat, labels = data
+
+    in_feats = nfeat.shape[1]
+
+    # Create PyTorch DataLoader for constructing blocks
+    n_edges = g.num_edges()
+    train_seeds = th.arange(n_edges)
+
+    if args.graph_device == "gpu":
+        train_seeds = train_seeds.to(device)
+        g = g.to(device)
+        args.num_workers = 0
+    elif args.graph_device == "uva":
+        train_seeds = train_seeds.to(device)
+        g.pin_memory_()
+        args.num_workers = 0
+
+    # Create sampler
+    sampler = dgl.dataloading.NeighborSampler(
+        [int(fanout) for fanout in args.fan_out.split(",")],
+        prefetch_node_feats=["features"],
+    )
+    sampler = dgl.dataloading.as_edge_prediction_sampler(
+        sampler,
+        exclude="reverse_id",
+        # For each edge with ID e in Reddit dataset, the reverse edge is e ± |E|/2.
+        reverse_eids=th.cat(
+            [th.arange(n_edges // 2, n_edges), th.arange(0, n_edges // 2)]
+        ).to(train_seeds),
+        negative_sampler=NegativeSampler(
+            g,
+            args.num_negs,
+            args.neg_share,
+            device if args.graph_device == "uva" else None,
+        ),
+    )
+    dataloader = dgl.dataloading.EdgeDataLoader(
+        g,
+        train_seeds,
+        sampler,
+        device=device,
+        use_ddp=n_gpus > 1,
+        batch_size=args.batch_size,
+        shuffle=True,
+        drop_last=False,
+        num_workers=args.num_workers,
+        use_uva=args.graph_device == "uva",
+    )
+
+    dataloader.attach_ndata(
+        "features", GPUCachedTensorStorage(nfeat, args.cache_size)
+    )
```

Comment:
The current implementation uses the active CUDA stream to fetch features from the GPU cache and also from RAM using UnifiedTensor through UVA access. Is there a way to overlap queries to the GPU cache and to the UnifiedTensor
to hide feature fetch latency? I would assume that only a few SMs can use the whole bandwidth of the PCI-e when fetching from UnifiedTensor so using another stream, it may be possible to hide this latency.

Reply:
Once PR https://github.com/dmlc/dgl/pull/4945 gets merged, the FeatureStorage attached to the dataloader will be fetched in another stream.
https://github.com/dmlc/dgl/blob/e0fc038db111dede9792b3a8158af21dfa4c0748/python/dgl/dataloading/dataloader.py#L333-L340
But for now, it won't overlap with training due to sequential CPU workloads. In an internal project, we make it by using another CPU thread for feature fetching. We will see if we can upstream the modification.